xla/service/gpu/transforms/windowed_einsum_handler.cc

```diff
@@ -492,6 +511,125 @@ absl::Status PostProcessUnrolledLoop(HloInstruction* loop, int64_t stream_id) {
+  //    \                     /                   /
+  //     \                   /                   /
+  //     shape(n,x,y) concatenate on first axis, n is the number of partitions
+  //                        |
+  //             shape(n,x,y) loop output
+  //                        |
+  //             shape(x,y) reduction on first axis
+  //
+  // The final reduction is pulled outside of the loop to overlap with other
+  // collectives.
+  if (partial_accumulations.size() > 0 &&
+      while_body->name().find(
+          WindowedEinsumHandler::kWindowedEinsumAgLoopName) !=
+          std::string::npos) {
+    std::vector<HloInstruction*> partials_to_concat;
+
+    // We reshape it to a N+1 dimensioned tensor with left-most dim being 1.
+    Shape shape = partial_accumulations[0]->shape();
+    shape = ShapeUtil::PrependMajorDimension(1, shape);
+
+    for (auto& inst : partial_accumulations) {
+      CHECK(inst->user_count() == 1);
+      HloInstruction* reshaped_partial = while_body->AddInstruction(
+          HloInstruction::CreateReshape(shape, inst));
+      partials_to_concat.push_back(reshaped_partial);
+    }
+    Shape concat_shape = partial_accumulations[0]->shape();
+    concat_shape = ShapeUtil::PrependMajorDimension(
+        partial_accumulations.size(), concat_shape);
+
+    HloInstruction* concat = while_body->AddInstruction(
+        HloInstruction::CreateConcatenate(concat_shape, partials_to_concat, 0));
+
+    HloComputation* comp = loop->parent();
+    HloInstruction* windowed_lhs = loop->mutable_operand(0)->mutable_operand(0);
+    // Add a broadcasted zero of the same type as windowed_lhs. This holds all
+    // the partial accumulations and will be fed to a global reduction after
+    // this windowed einsum loop. We move the reduction outside of the loop so
+    // it can be fused or overlap with other instructions in the main
+    // computation.
+    Literal zero_literal =
+        LiteralUtil::Zero(windowed_lhs->shape().element_type());
+    HloInstruction* zero = comp->AddInstruction(
+        HloInstruction::CreateConstant(std::move(zero_literal)));
+    Shape zero_bcast_shape = ShapeUtil::ChangeElementType(
+        concat_shape, windowed_lhs->shape().element_type());
+    HloInstruction* zero_bcast = MakeBroadcastHlo(zero, {}, zero_bcast_shape);
+    loop->mutable_operand(0)->AppendOperand(zero_bcast);
+    ShapeUtil::AppendShapeToTuple(zero_bcast->shape(),
+                                  loop->mutable_operand(0)->mutable_shape());
+
+    // Update the parameter tuples of while's body and condition
+    // computations.
+    for (HloComputation* while_comp : {while_body, loop->while_condition()}) {
+      while_comp->ReplaceParameter(
+          0, HloInstruction::CreateParameter(
+                 0, loop->mutable_operand(0)->shape(),
+                 while_comp->parameter_instruction(0)->name()));
+    }
+    HloInstruction* root = while_body->root_instruction();
```

Comment:
Instead of creating a new root instruction can you just mute the original root instruction? Similar to how you mutated the operand on the line `loop->mutable_operand(0)->AppendOperand(zero_bcast);`?

Reply:
For some reason when i tried just mutating the instruction directly, it didnt change the call signature of the while computation so the verifier failed. I think that might work for a single instruction but not for root instruction of a comp. Replacing the instruction is the safest way it seems.